mixer/pkg/protobuf/yaml/dynamic/auth.go

```diff
@@ -0,0 +1,234 @@
+		ClientID:     oauthCfg.ClientId,
+		ClientSecret: string(secret),
+		TokenURL:     oauthCfg.TokenUrl,
+		Scopes:       oauthCfg.Scopes,
+	}
+	return c.TokenSource(context.Background()), nil
+}
+
+func buildAuthToken(token []byte, tlsCfg *policypb.Tls) (*headerToken, error) {
+	if tlsCfg.TokenType == nil {
+		return nil, errors.New("token type should be specified")
+	}
+	at := &headerToken{
+		token: token,
+	}
+	switch tlsCfg.TokenType.(type) {
+	case *policypb.Tls_AuthHeader_:
+		ah := tlsCfg.GetAuthHeader()
+		at.header = "authorization"
+		if ah == policypb.PLAIN {
+			at.prefix = ""
+		} else if ah == policypb.BEARER {
+			at.prefix = "Bearer"
+		}
+	case *policypb.Tls_CustomHeader:
+		at.prefix = ""
+		at.header = tlsCfg.GetCustomHeader()
+	default:
+		return nil, errors.New("authToken type cannot be empty")
+	}
+	return at, nil
+}
+
+func buildPerRPCCredentials(tlsCfg *policypb.Tls) (credentials.PerRPCCredentials, error) {
+	switch tlsCfg.TokenSource.(type) {
+	case *policypb.Tls_Oauth:
+		ts, err := buildOAuthToken(tlsCfg.GetOauth())
+		if err != nil {
+			return nil, err
+		}
+		// nolint: govet
+		return oauth.TokenSource{ts}, nil
+	case *policypb.Tls_TokenPath:
+		// read token from given path
+		tp := tlsCfg.GetTokenPath()
+		if tp == "" {
+			return nil, errors.New("token path should not be empty")
+		}
+		token, err := ioutil.ReadFile(tp)
+		if err != nil {
+			return nil, err
+		}
+
+		// initialize per rpc credential based on token type
+		ht, err := buildAuthToken(token, tlsCfg)
+		return ht, err
+	default:
+		return nil, errors.New("unexpected tls token source type")
+	}
+}
+
+func buildTLSDialOption(tlsCfg *policypb.Tls, skipVerify bool) ([]grpc.DialOption, error) {
+	var opts []grpc.DialOption
+	// Load certificates, include system certs, istio certs and additional certs from config.
+	caCertPool, err := loadCerts(tlsCfg.GetCaCertificates())
+	if err != nil {
+		return nil, fmt.Errorf("failed build tls dial option: ca cert cannot be load %v", err)
+	}
+
+	tc := credentials.NewTLS(&tls.Config{
+		RootCAs:            caCertPool,
+		InsecureSkipVerify: skipVerify,
+	})
+	opts = append(opts, grpc.WithTransportCredentials(tc))
+
+	// Load per RPC credential.
+	perRPCCreds, err := buildPerRPCCredentials(tlsCfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed build tls dial option: cannot get grpc per rpc credentials %v", err)
+	}
+	opts = append(opts, grpc.WithPerRPCCredentials(perRPCCreds))
+	return opts, nil
+}
+
+func (a *authHelper) getAuthOpt() (opts []grpc.DialOption, err error) {
+	opts = []grpc.DialOption{}
+	if a.authCfg == nil {
+		opts = append(opts, grpc.WithInsecure())
+		return opts, nil
+	}
+	switch t := a.authCfg.AuthType.(type) {
+	case *policypb.Authentication_Tls:
+		opts, err = buildTLSDialOption(a.authCfg.GetTls(), a.skipVerification)
+		if err != nil {
+			return nil, err
+		}
+	case *policypb.Authentication_Mutual:
+		dialAuthOpt, err := buildMTLSDialOption(a.authCfg.GetMutual(), a.skipVerification)
```

Comment:
This seems convoluted. why not just `
return `buildMTLSDialOption(a.authCfg.GetMutual(), a.skipVerification)` and similar everywhere?